### PR TITLE
Layout Cycle Crash + Misc Fixes for NavigationView

### PR DESCRIPTION
**Pull request by karenbtlai**
## Description

**Layout Cycle Crash Fix:**

`UpdatePaneLayout()` was calculating Separator and Footer height from hardcoded values. The discrepancy from the template was causing ScrollViewer to continuously calculate incrementally different heights and crashing. 

It is now refactored to include all footer group items (`footerItemsRepeater` `paneFooter`, `VisualItemsSeparator`) and its margins top and bottom.
 
**Miscellaneous fixes**
- Update brushes
- Expose certain themeresources
- Fix separator margins

### CODE REVIEW

File: dev/NavigationView/NavigationView.xaml [hunk 1]
```diff
@@ -927,7 +930,7 @@
                             x:Name="SeparatorLine"
                             Height="{ThemeResource NavigationViewItemSeparatorHeight}"
                             Fill="{ThemeResource NavigationViewItemSeparatorForeground}"
-                            Margin="{TemplateBinding Margin}"/>
+                            Margin="{ThemeResource NavigationViewItemSeparatorMargin}"/>
```
Comment (beervoley) [inline dev/NavigationView/NavigationView.xaml:933]: what is the reason for this change?:)

File: dev/NavigationView/TestUI/Common/NavigationViewPage.xaml [hunk 2]
```diff
@@ -26,6 +26,7 @@
                 DisplayModeChanged="NavView_DisplayModeChanged"
                 ItemInvoked="NavView_ItemInvoked"
                 SelectionChanged="NavView_SelectionChanged"
+                Header="TestHeader"
```
Comment (beervoley) [inline dev/NavigationView/TestUI/Common/NavigationViewPage.xaml:29]: what is this change for ?:)

File: dev/NavigationView/NavigationView.xaml (hunk 1 above)
Comment (karenbtlai) [reply dev/NavigationView/NavigationView.xaml:933]: The template binding was causing the margins to be set twice. This sets the properly defined margins on the separator.

File: dev/NavigationView/TestUI/Common/NavigationViewPage.xaml (hunk 2 above)
Comment (karenbtlai) [reply dev/NavigationView/TestUI/Common/NavigationViewPage.xaml:29]: I guess I was using this to see where the header goes. Helpful, so I decided to keep it 😂

File: dev/NavigationView/NavigationView_rs1_themeresources.xaml [hunk 3]
```diff
@@ -211,7 +215,9 @@
 
             <SolidColorBrush x:Key="NavigationViewSelectionIndicatorForeground" Color="{ThemeResource SystemColorHighlightTextColor}" />
 
-            <StaticResource x:Key="TopNavigationViewItemForeground" ResourceKey="SystemControlForegroundBaseHighBrush" />
+            <StaticResource x:Key="NavigationViewContentGridBorderBrush" ResourceKey="SubtleFillColorTransparentBrush" />
```
Comment (beervoley) [inline dev/NavigationView/NavigationView_rs1_themeresources.xaml:218]: it's HC, are we supposed to user other brushes than **System***?:)

**Review by gabbybilka — APPROVED** [review]

File: dev/NavigationView/NavigationView_rs1_themeresources.xaml (hunk 3 above)
Comment (karenbtlai) [reply dev/NavigationView/NavigationView_rs1_themeresources.xaml:218]: All the NavViewItemBorderBrush for HC is using the same brush. But you're right, maybe it should be `SystemControlTransparentBrush` for consistency

**Review by beervoley — APPROVED** [review]

**Review by chingucoding — APPROVED** [review]
LGTM

File: dev/NavigationView/NavigationView.h [hunk 4]
```diff
@@ -361,6 +361,7 @@ class NavigationView :
     tracker_ref<winrt::Border> m_topNavContentOverlayAreaGrid{ this };
     tracker_ref<winrt::Grid> m_shadowCaster{ this };
     tracker_ref<winrt::Storyboard> m_shadowCasterEaseOutStoryboard{ this };
+    tracker_ref<winrt::NavigationViewItemSeparator> m_visualItemsSeparator{ this };
```
Comment (RBrid) [inline dev/NavigationView/NavigationView.h:364]: Does not look like you use this anymore.

File: dev/NavigationView/NavigationView.cpp [hunk 5]
```diff
@@ -85,6 +85,7 @@ static constexpr auto c_paneHeaderContentBorderRow = L"PaneHeaderContentBorderRo
 
 static constexpr auto c_separatorVisibleStateName = L"SeparatorVisible";
 static constexpr auto c_separatorCollapsedStateName = L"SeparatorCollapsed";
+static constexpr auto c_visualItemsSeparator = L"VisualItemsSeparator"sv;
```
Comment (RBrid) [inline dev/NavigationView/NavigationView.cpp:88]: Does not look like you use this anymore.

**Review by RBrid — CHANGES_REQUESTED** [review]

File: dev/NavigationView/NavigationView.cpp [hunk 6]
```diff
@@ -1533,6 +1514,17 @@ void NavigationView::UpdatePaneLayout()
                             const auto footersActualHeight = footerItemsRepeater.ActualHeight();
                             const auto menuItemsActualHeight = menuItems.ActualHeight();
 
+                            const auto paneFooterHeight = [this]() {
```
Comment (RBrid) [inline dev/NavigationView/NavigationView.cpp:1517]: For consistency, this should be called paneFooterActualHeight.

So this line will look better:
const auto footerGroupActualHeight = footersActualHeight + paneFooterHeight;
-->
const auto footerGroupActualHeight = footersActualHeight + paneFooterActualHeight;

File: dev/NavigationView/NavigationView.cpp [hunk 7]
```diff
@@ -1533,6 +1514,17 @@ void NavigationView::UpdatePaneLayout()
...
+                                if (const auto& paneFooter = m_leftNavFooterContentBorder.get())
+                                {
+                                    return paneFooter.ActualHeight();
```
Comment (RBrid) [inline dev/NavigationView/NavigationView.cpp:1520]: ActualHeight does not include Margin.Top and Margin.Bottom.  That's for instance a problem in the first MUXControlsTestApp NavigationView test page where paneFooter.Margin.Bottom is 4px (probably the new default).

You need to use 
const auto paneFooterMargin = paneFooter.Margin();

Then 
return paneFooter.ActualHeight() + paneFooterMargin.Top + paneFooterMargin.Bottom;

File: dev/NavigationView/NavigationView.cpp [hunk 8]
```diff
@@ -1544,12 +1536,12 @@ void NavigationView::UpdatePaneLayout()
                                 winrt::VisualStateManager::GoToState(*this, c_separatorCollapsedStateName, false);
                                 return 0.0;
                             }
-                            else if (totalAvailableHeight > menuItemsActualHeight + footersActualHeight)
+                            else if (totalAvailableHeight > menuItemsActualHeight + footerGroupActualHeight)
```
Comment (RBrid) [inline dev/NavigationView/NavigationView.cpp:1539]: In that MUXControlsTestPage first NavigationView test page, totalAvailableHeight happens to be menuItemsActualHeight + footerGroupActualHeight + 4.0 

The 4.0 comes from the missing paneFooter.Margin.Bottom mentioned above. Once that is fixed, you will end up with  
totalAvailableHeight == menuItemsActualHeight + footerGroupActualHeight and this branch will incorrectly be skipped.

I believe this is because instead of:
else if (totalAvailableHeight > menuItemsActualHeight + footerGroupActualHeight)
we should be using:
else if (totalAvailableHeight >= menuItems.DesiredSize.Height + footerGroupActualHeight)

File: dev/NavigationView/NavigationView.cpp [hunk 9]
```diff
@@ -1533,6 +1514,17 @@ void NavigationView::UpdatePaneLayout()
                             const auto footersActualHeight = footerItemsRepeater.ActualHeight();
```
Comment (RBrid) [inline dev/NavigationView/NavigationView.cpp:1514]: footerItemsRepeater.Margin.Top + footerItemsRepeater.Margin.Bottom happens to be 0, but they should be added in case margins are set.

File: dev/NavigationView/NavigationView.cpp [hunk 10]
```diff
@@ -1533,6 +1514,17 @@ void NavigationView::UpdatePaneLayout()
...
                             const auto menuItemsActualHeight = menuItems.ActualHeight();
```
Comment (RBrid) [inline dev/NavigationView/NavigationView.cpp:1515]: menuItems.Margin.Top + menuItems.Margin.Bottom happens to be 0, but they should be added in case margins are set.

File: dev/NavigationView/NavigationView.cpp (hunk 8 above)
Comment (karenbtlai) [reply dev/NavigationView/NavigationView.cpp:1539]: I don't think this is being incorrectly skipped. If `totalAvailableHeight == menuItemsActualHeight + footerGroupActualHeight`, logically, there should be a pane separator to distinguish the two sections rather than it being collapsed.

File: dev/NavigationView/NavigationView.cpp [hunk 11]
```diff
@@ -1530,8 +1512,36 @@ void NavigationView::UpdatePaneLayout()
...
+                                if (const auto& visualItemsSeparator = m_visualItemsSeparator.get())
+                                {
+                                    const auto visualItemsSeparatorMargin = visualItemsSeparator.Margin();
+                                    return visualItemsSeparator.ActualHeight() + visualItemsSeparatorMargin.Top + visualItemsSeparatorMargin.Bottom;
+                                }
+                                return 0.0;
+                            }();
+
+                            // Footer, PaneFooter, and VisualItemsSeparator are included in the footerGroup to calculate available height for menu items.
+                            const auto footerGroupActualHeight = footersActualHeight + paneFooterActualHeight + visualItemsSeparatorHeight;
```
Comment (RBrid) [inline dev/NavigationView/NavigationView.cpp:1544]: Adding the visualItemsSeparatorHeight feels incorrect to me:  you don't want the separator to be shown because of a lack of space due to the separator being shown. You know what I mean?

Comment (karenbtlai) [reply dev/NavigationView/NavigationView.cpp:1544]: I thought about that too... but it was part of the calculation before so I simply included it.

Comment (karenbtlai) [reply dev/NavigationView/NavigationView.cpp:1544]: Discussed offline, will remove.

File: dev/NavigationView/NavigationView.cpp [hunk 12]
```diff
@@ -1513,35 +1511,48 @@ void NavigationView::UpdatePaneLayout()
...
+                                    return menuItemsMargin.Top + menuItemsMargin.Bottom;
                                 }
                                 return 0.0;
                             }();
 
-                            // Footer, PaneFooter, and VisualItemsSeparator are included in the footerGroup to calculate available height for menu items.
-                            const auto footerGroupActualHeight = footersActualHeight + paneFooterActualHeight + visualItemsSeparatorHeight;
+                            // This is the value computed during the measure pass of the layout process. This will be the value used to determine
+                            // the partition logic between menuItems and footerGroup, since the ActualHeight may be taller if there's more space.
+                            const auto menuItemsDesiredHeight = menuItems.DesiredSize().Height + menuItemsTopBottomMargin;
```
Comment (RBrid) [inline dev/NavigationView/NavigationView.cpp:1548]: Unlike ActualHeight/ActualWidth, UIElement.DesiredSize already includes the Margin. So it should be:

`const auto menuItemsDesiredHeight = menuItems.DesiredSize().Height;`

**Review by RBrid — CHANGES_REQUESTED** [review]

**Review by RBrid — APPROVED** [review]

File: dev/NavigationView/NavigationView.xaml [hunk 13]
```diff
@@ -181,10 +185,10 @@
                             </VisualStateGroup>
 
                             <VisualStateGroup x:Name="PaneSeparatorStates">
-                                <VisualState x:Name="SeparatorCollapsed"/>
-                                <VisualState x:Name="SeparatorVisible">
+                                <VisualState x:Name="SeparatorVisible"/>
+                                <VisualState x:Name="SeparatorCollapsed">
                                     <VisualState.Setters>
-                                        <Setter Target="VisualItemsSeparator.Visibility" Value="Visible"/>
+                                        <Setter Target="VisualItemsSeparator.Visibility" Value="Collapsed"/>
```
Comment (chingucoding) [inline dev/NavigationView/NavigationView.xaml:191]: Out of curiosity, why is this switched now?

**Review by chingucoding — APPROVED** [review]

File: dev/NavigationView/NavigationView.cpp (hunk 8 above)
Comment (karenbtlai) [reply dev/NavigationView/NavigationView.cpp]: Discussed offline. You're right, it would make more sense to add the >=

File: dev/NavigationView/NavigationView.xaml (hunk 13 above)
Comment (karenbtlai) [reply dev/NavigationView/NavigationView.xaml:191]: oh oops. I was previously getting the separator ActualHeight from codebehind, and for it to evaluate to a non-zero value, it needed to be visible by default. The separator is no longer used for pane layout calculations, so I should revert this.
